### ISSUE TITLE
Redirect to original site when first confirming e-mail or change wording

### ISSUE THREAD

**Issue by wolftune** [mozilla/persona#2718]
Ok, this is a confusion for new users. Here's the situation:
1. Go to site that uses Persona having never used it before
2. Click log in and choose to sign up with Persona
3. Get this message: "Check your email at [email]. Click the link in the confirmation email. You'll then immediately be signed in to [site].
4. **\* there is no specific instruction on how to go to your e-mail, such as opening it in a new tab versus just going there
5. E-mail says "Thank you for using Persona. Click to confirm this email address and automatically sign in to [site]"
6. After clicking to confirm, the confirmation site says: "Thank you for signing up with Persona. You can now use your Persona account to Sign In or Sign Up to websites all across the web! Your new address is set up and ready to go. You will be redirected to https://login.persona.org/ Redirecting in 1 second"
7. The site redirects to login.persona.org

So naive users go through this process and then are sitting there at an unfamiliar site with no instructions about what to do. There's this expectation that users know to return to the site they wanted, but this isn't how most sites work. After confirming an account, users expect to be back at the site they wanted to go to. This is especially confusing if they didn't open a new tab to go to their e-mail.

So there needs to be a way to at least instruct people to manually go back to the site they were at. Better would be to have the redirect go to the site from which they got the initial login that led to signing up, or at least a link to that site.

I mention this because I actively dealt with a user of my site asking for help and being confused. She clicked what she thought would sign into my site and ended up at this other site and just gave up and felt confused.

**Comment by shane-tomlinson** [comment]
@wolftune - Thanks for the feedback! I think we already have your case covered, but you must be using .watch & .request instead of .get [1], is this the case? This can also happen if you open the verification link in a browser or device that is different to the browser/device you started the sign in process on. 

The tab the user opens the email in should not matter.

If neither of these are the case, we have a deeper problem - can you let us know?

# 

[1] - https://developer.mozilla.org/en-US/docs/DOM/navigator.id

**Comment by wolftune** [comment]
I think you may have misunderstood. Everything is completely successful in terms of being logged in to the site.

The problem is that user goes through steps and doesn't end up specifically looking at the intended site. The user has to either A. go back to the tab which has the original site loaded, or B. click the link or enter the URL of the original site again to be there.

This may seem obvious to some of us, but not to everyone.
Say I close the original site after signing up with Persona. I go through the e-mail confirmation etc. At no point does Persona put me back on the original site or simply instruct me "now you may go back to the original site"
Instead, it can seem from the text like everything will happen automatically, but instead I end up looking at the Persona site.
So for really naive computer users, this is frustrating.

I want a way for the e-mail confirmation to actually redirect to the site where Persona was first initiated OR I want there to be very clear text instructions: "now you may close this page and return to the original site, if it is still open in a tab or window, simply go to that, or you may load the original site via your original link or URL address"

In other words: really explicit web-for-dummies type of concern here.

Am I clear? Sorry if you understood originally, but I wasn't sure about it based on your reply.

**Comment by shane-tomlinson** [comment]
@wolftune 

> The problem is that user goes through steps and doesn't end up specifically looking at the intended site. 
> The user has to either A. go back to the tab which has the original site loaded, or B. click the link or enter
>  the URL of the original site again to be there.

This is the part that I am confused about. Persona _should_ automatically redirect you back to the original site after you click on the verification email, with the caveat that the site _must_ be using the .watch/.request API and not the .get API.

**Comment by shane-tomlinson** [comment]
@wolftune - you can try the expected flow from our demo site http://123done.org, sign up as a new user. After you click the verification link in the email, it should open the link on persona.org and then redirect you back to http://123done.org. After you land on 123done.org, it should sign you in automatically. If this is not happening, would you be willing to help us debug why? This is pretty core to our expected behavior and a regression here warrants deeper investigation.

**Comment by wolftune** [comment]
Thanks, Shane!

You were right. It turns out that we were using the .get. The thing is, that wasn't a conscious choice, it was in the internals of the library we were using.

However, although we've now succeeded in getting out site working, there remains a problem that should be fixed with .get:
The email the users receive makes it sound like they will be forwarded to the site and instead they are forwarded to the persona account page. Since persona knows where they will be sent when it generates the email, the wording should be changed even if the functionality is not, to make this clearer.

I understand that technically "you will be automatically logged in" does not mean "you will be redirected" , but most people will assume it. So it is not good for Persona in general to have the redirect to Persona instead of the site given this wording. So I urge you to at least change the wording.

Again, our site is now fixed with your other suggestions, but we want Persona to be less confusing in general, even for other sites that continue to use .get alone.

Cheers,
Aaron

**Comment by shane-tomlinson** [comment]
@wolftune - Thanks for checking the underlying code in your library, I am glad that this is resolved for you. 

I completely understand what you are getting at with the email, there was talk at yesterday's bug scrub that the email text was already changed for this exact reason, it is just in the queue to be delivered to the production servers.

This was done in aec53d9cd8e7ff8dd494ad76af696dd17f4840dd.

I'm not sure if this addresses your request, any additional feedback is appreciated!

**Comment by wolftune** [comment]
That doesn't fix it, it just helps the wording be more concise.
To fix the issue here without technical changes (i.e by just changing the text), it needs to say something like:
"Click to confirm this email address. This will then send you to the Persona account page. Afterward, when you manually return to %s, you will be automatically logged in."

I wish it could be less words but it needs to be clear somewhere that the user has to take some action to get back to %s. Perhaps the Persona website where .get redirects to could include a prominent link to %s along with the other welcome and account stuff?

The bigger question is: why not incorporate .watch & .request within .get? I.e what is the purpose of having a setup which does _not_ redirect to %s?

EDIT: I guess if .get is really deprecated so nobody uses it anymore, then this isn't needed… I'm not sure here though

**Comment by ozten** [comment]
> The thing is, that wasn't a conscious choice, it was in the internals of the library we were using.

Can you file a bug against that library to upgrade to the `watch` API? `get` is deprecated. Which one is it, out of curiosity?

**Comment by wolftune** [comment]
Sure, I'll look into that. It was the BrowserID plugin for Yesod (http://www.yesodweb.com/)

See https://github.com/yesodweb/yesod/blob/master/yesod-auth/Yesod/Auth/BrowserId.hs

I opened this ticket:
https://github.com/yesodweb/yesod/issues/453

Thanks for the help, and sorry for my confusion. Just trying to help everything be clear.

**Comment by snoyberg** [comment]
Author of the plugin in question. I've updated my code, but in the process I ran into two quirks.
1. It seems like the only reason I need to switch to the .watch API to address this issue is to be able to provide the returnTo parameter. Is there any inherent reason not to allow that parameter in .get?
2. The Yesod authentication system is built to allow multiple authentication systems, including OpenID. Suppose that a user is logged in via OpenID; there's no way to indicate via a system besides the Mozilla Persona-recognized email address. As such, onlogin can be called even if the user is logged into the site via another means.

I consider the second issue more important. I worked around it by only calling watch after the user clicks on "Login", but this seems like a bit of a hack. Is there any recommended practice for how to deal with a scenario where users can log into a site via multiple means? (If it would be better, I can open this as a separate issue.)

**Comment by callahad** [comment]
@snoyberg A key feature of `returnTo` is that, when the user returns, they're logged in on that site. That's only possible with `watch`, since it lets us automatically fire an `onlogin` action when the user lands on the page. We couldn't do that with `get`, so we haven't exposed it there.

For the second issue, could you write to our mailing list? It seems like there could be room for additional discussion there... but in short, `onlogin` will only fire on its own if the user has manually elected to log into the site with Persona, and is returning later. So it seems like it should Just Work for you, even if there are multiple authentication methods on the page.

(Edit: mailing list is at https://lists.mozilla.org/listinfo/dev-identity)

**Comment by snoyberg** [comment]
Thanks for the response. I hadn't realized that `onlogin` fired that way, it might work out perfectly for me. I'll try and test it out tomorrow and let you know the results.

**Comment by callahad** [comment]
Hi! To help us better focus, I'm "closing" all issues that have been open for more than six months. These have been tagged "cleanup-2014" so that we can go back and review them in the future.

For more information, check out this thread: http://thread.gmane.org/gmane.comp.mozilla.identity.devel/7394

If you believe this bug is still a major issue for you, please comment, submit a pull request, or discuss it on our mailing list: https://lists.mozilla.org/listinfo/dev-identity

Sorry for GitHub notification spam!
